### PR TITLE
added backend implementation for model templating

### PR DESCRIPTION
**Pull request by ARADDCC002**

### CODE REVIEW

**Review by GB27247 — CHANGES_REQUESTED** [review]

File: backend/src/services/model.ts [hunk 1]
```diff
@@ -347,9 +346,28 @@ export async function createModelCardFromSchema(
...
-  _schemaId: string,
+  user: UserInterface,
+  modelId: string,
+  templateId: string,
 ): Promise<ModelCardRevisionDoc> {
-  throw NotImplemented({}, 'This feature is not yet implemented')
+  const model = await getModelById(user, modelId)
+  checkModelRestriction(model)
+  const template = await getModelById(user, templateId)
+  checkModelRestriction(template)
```
[the remark this answers is not in the document]
Comment (ARADDCC002) [reply backend/src/services/model.ts:356]: I think it's necessary to check that new model is not a mirrored model - there could be an instance that someone makes a mirrored model but doesn't create a model card, then tries to base it off a template.

As we are not changing the metadata of the template, however, I think we can safely remove this check.

File: backend/src/services/model.ts [hunk 2]
```diff
@@ -347,9 +346,28 @@ export async function createModelCardFromSchema(
...
+  const modelAuth = await authorisation.model(user, model, ModelAction.Write)
+  if (!modelAuth.success) {
+    throw Forbidden(modelAuth.info, { userDn: user.dn, modelId })
+  }
+  const templateAuth = await authorisation.model(user, template, ModelAction.View)
+  if (!templateAuth.success) {
+    throw Forbidden(templateAuth.info, { userDn: user.dn, modelId })
+  }
+
+  if (!template.card) {
```
[the remark this answers is not in the document]
Comment (ARADDCC002) [reply backend/src/services/model.ts:367]: When you first create a model `card` doesn't exist - I'm not sure where you're seeing it being set as an empty object, but if it is somehow being set let me know!

File: backend/src/services/model.ts [hunk 3]
```diff
@@ -347,9 +346,27 @@ export async function createModelCardFromSchema(
...
+  user: UserInterface,
+  modelId: string,
+  templateId: string,
 ): Promise<ModelCardRevisionDoc> {
-  throw NotImplemented({}, 'This feature is not yet implemented')
+  if (modelId === templateId) {
+    throw BadReq('The model and template ID must be different', { modelId, templateId })
+  }
+  const model = await getModelById(user, modelId)
+  checkModelRestriction(model)
```
Comment (ARADDCC012) [inline backend/src/services/model.ts:357]: Do we need to `checkModelRestriction` for the model being used as a template also?

File: backend/src/services/model.ts [hunk 4]
```diff
@@ -347,9 +346,27 @@ export async function createModelCardFromSchema(
...
+  const template = await getModelById(user, templateId)
+
+  // Check to make sure user can access the template. We already check for the model auth later on in _setModelCard
+  const templateAuth = await authorisation.model(user, template, ModelAction.View)
+  if (!templateAuth.success) {
+    throw Forbidden(templateAuth.info, { userDn: user.dn, modelId })
```
Comment (ARADDCC012) [inline backend/src/services/model.ts:363]: Shouldn't this be reporting the `templateId`?
```
throw Forbidden(templateAuth.info, { userDn: user.dn, modelId: templateId })
```

File: backend/src/services/model.ts [hunk 5]
```diff
@@ -347,9 +346,27 @@ export async function createModelCardFromSchema(
...
-  _user: UserInterface,
-  _modelId: string,
-  _schemaId: string,
+  user: UserInterface,
+  modelId: string,
+  templateId: string,
 ): Promise<ModelCardRevisionDoc> {
-  throw NotImplemented({}, 'This feature is not yet implemented')
+  if (modelId === templateId) {
+    throw BadReq('The model and template ID must be different', { modelId, templateId })
```
Comment (ARADDCC012) [inline backend/src/services/model.ts:354]: We should have a unit test that covers this

File: backend/src/services/model.ts (hunk 4 above)
Comment (ARADDCC012) [inline backend/src/services/model.ts:363]: We should have a unit test that covers this

File: backend/test/services/model.spec.ts [hunk 6]
```diff
@@ -311,4 +312,17 @@ describe('services > model', () => {
     )
     expect(modelMocks.save).not.toBeCalled()
   })
+
+  test('crateModelCardFromTemplate > requesting to use a template without a model card will throw an error', async () => {
```
Comment (ARADDCC012) [inline backend/test/services/model.spec.ts:316]: We should also have a "happy day" test

**Review by ARADDCC012 — CHANGES_REQUESTED** [review]

File: backend/src/services/model.ts (hunk 3 above)
Comment (ARADDCC002) [reply backend/src/services/model.ts:357]: I don't think so. All it does is check to see if it's "mirrored" which is basically read only. As we're not editing the template in any way, I think it's fine to use it if it's read only.

File: backend/src/services/model.ts [hunk 7]
```diff
@@ -347,9 +346,25 @@ export async function createModelCardFromSchema(
...
+  // Check to make sure user can access the template. We already check for the model auth later on in _setModelCard
+  const templateAuth = await authorisation.model(user, template, ModelAction.View)
+  if (!templateAuth.success) {
+    throw Forbidden(templateAuth.info, { userDn: user.dn, templateId })
+  }
+
+  if (!template.card?.schemaId) {
+    throw BadReq('The template model is missing a model card', { modelId, templateId })
+  }
+  const revision = await _setModelCard(user, modelId, template.card.schemaId, 1, template.card.metadata)
```
Comment (JR40159) [inline backend/src/services/model.ts:368]: Is there anything to stop a user from calling this with a `modelId` that references a model that already has model card versions? If this were to happen then this line would set the first version to be the model card of the template and all later versions would be from the pre-existing model card.

**Review by JR40159 — CHANGES_REQUESTED** [review]

**Review by JR40159 — APPROVED** [review]

**Pull request merged**